### PR TITLE
reserve the status for operators which are still in deletion stage

### PR DESCRIPTION
**Pull request by Daniel-Fan**
### Background

ODLM get the operators which will be deleted by [comparing the differences](https://github.com/IBM/operand-deployment-lifecycle-manager/blob/2e39cfbddb9b3a3b7092e12fac37947ce2777026/controllers/operandrequest/reconcile_operator.go#L430) between operators in `OperandRequest` status with operators in `OperandRequest` spec.

### Root cause
At the end of each reconciliation, ODLM will [refresh the `OperandRequest` status](https://github.com/IBM/operand-deployment-lifecycle-manager/blob/2e39cfbddb9b3a3b7092e12fac37947ce2777026/controllers/operandrequest/reconcile_operator.go#L52) based on existing operators in `OperandRequest` status.
When we remove an operator from `OperandRequest` spec, it will be removed from `OperandRequest` status as well.

However, if the deletion of those operators hits an error, those operators will be left in the cluster.
But ODLM will still refresh the status of `OperandRequest` based on spec in current and following reconciliation. And as a consequence, the status of those operators left in the cluster is gone.

ODLM will never find the information about those orphan operators in the comparison of status and spec, and continue the deletion process.

### Solution
When ODLM is refreshing the status of OperandRequest. It will go through all operators in existing status.
The status of operators will be reserved if they are either in the `OperandRequest` spec or exist in cluster(Subscription exists).

#### Reproduce the issue
1. Create a OperandRequest containing operator `ibm-cert-manager-operator` and `cloud-native-postgresql`
```
apiVersion: operator.ibm.com/v1alpha1
kind: OperandRequest
metadata:
  name: example-service
  namespace: cp4d
spec:
  requests:
    - operands:
        - name: cloud-native-postgresql
        - name: ibm-cert-manager-operator
      registry: common-service
```

2. Check the status of `OperandRequest`, both `ibm-cert-manager-operator` and `cloud-native-postgresql` exists.
```
  members:
    - name: cloud-native-postgresql
      phase:
        operandPhase: Running
        operatorPhase: Running
    - name: ibm-cert-manager-operator
      phase:
        operandPhase: Running
        operatorPhase: Running
  phase: Running
```

3. Add a finalizer in cert-manager CR
```
apiVersion: operator.ibm.com/v1alpha1
kind: CertManager
metadata:
  finalizers:
    - not-uninstall
```
4. Remove `ibm-cert-manager-oeprator` from `OperandRequest`, keep `cloud-native-posygresql` unchanged
5. Wait to see the error msg from ODLM pod logs, it happens because the ODLM is not able to delete cert-manager CR due to the finalizer we added into before.
```
E1128 00:11:46.223735 1 operandrequest_controller.go:92] failed to clean up the subscriptions for OperandRequest cp4d/example-service: the following errors occurred:
- the following errors occurred:
- failed to delete custom resource -- Kind: CertManager, NamespacedName: cs-control/default: timed out waiting for the condition
```
6. OperandRequest do not has the status for `ibm-cert-manager-operator` anymore, but `ibm-cert-manager-operator` exists in the cluster.
```
  members:
    - name: cloud-native-postgresql
      phase:
        operandPhase: Running
        operatorPhase: Running
  phase: Running
```
7. We manually remove finalizer in Cert-Manager CR, Cert-Manager CR is gone automatically. But `ibm-cert-manager-operator`  remains in the cluster, ODLM does not delete it anymore.

#### Patch ODLM image in CSV: `quay.io/daniel_fan/odlm:3664b25d-dirty`
1. Add `ibm-cert-manager-operator` back into the OperandRequest spec
```
spec:
  requests:
    - operands:
        - name: cloud-native-postgresql
        - name: ibm-cert-manager-operator
```
2. Add finalizer into Cert-Manager CR as what we previously did
3. Remove `ibm-cert-manager-oeprator` from `OperandRequest`, keep `cloud-native-posygresql` unchanged
4. Wait to see the error msg from ODLM pod logs, it happens because the ODLM is not able to delete cert-manager CR due to the finalizer we added into before.
```
E1128 00:11:46.223735 1 operandrequest_controller.go:92] failed to clean up the subscriptions for OperandRequest cp4d/example-service: the following errors occurred:
- the following errors occurred:
- failed to delete custom resource -- Kind: CertManager, NamespacedName: cs-control/default: timed out waiting for the condition
```
5. OperandRequest still HAS the status for `ibm-cert-manager-operator`.
```
  members:
    - name: cloud-native-postgresql
      phase:
        operandPhase: Running
        operatorPhase: Running
    - name: ibm-cert-manager-operator
      phase:
        operandPhase: Running
        operatorPhase: Running
  phase: Running
```
6. We manually remove finalizer in Cert-Manager CR, Cert-Manager CR is gone. ODLM will continue to uninstall `ibm-cert-manager-operator`. And finally `ibm-cert-manager-operator` will be uninstalled.

Signed-off-by: Daniel Fan <fanyuchensx@gmail.com>

### CODE REVIEW

**Review by bitscuit — CHANGES_REQUESTED** [review]
Verified fix works, but review readability comments for the code and also error handling

File: controllers/operandrequest/reconcile_operator.go [hunk 1]
```diff
@@ -549,3 +551,42 @@ func (r *Reconciler) checkUninstallLabel(ctx context.Context, name, namespace st
...
+								existingOperands = append(existingOperands, m.Name)
+								returnErr = fmt.Errorf("failed to delete operator %s in the namespace %s", m.Name, namespace)
+								break
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Daniel-Fan) [reply controllers/operandrequest/reconcile_operator.go]: Thank you for pointing this out, Henry!

I updated the code, please have a look at it when you are available.

**Comment by Daniel-Fan** [comment]
@bitscuit 

I worked out a more elegant way to resolve the issue. Please review it when you are available. Thank you!

**Review by bitscuit — APPROVED** [review]
/lgtm

**Pull request merged**